Ticket #29340 - Fix FileNotFoundError and add fenced_code extension

rastytheamateur opened the pull request.

Commit:
Ticket #29340 - Fix FileNotFoundError and add fenced_code extension

jiri-janousek reviewed (approved):
LGTM.

zbohm commented:
In my opinion, the check that the file exists should also take place during the validation of the admin form.

zbohm commented:
The same check should be performed on the repository URL.

rastytheamateur commented:
> In my opinion, the check that the file exists should also take place during the validation of the admin form.

Disagree, a file can be prepared in a separated branch for merge as well as web can be prepared for its publishing.

> The same check should be performed on the repository URL.

Agree, I am going to fix this.

jiri-janousek commented:
Zděněk wrote something about a prompt for a password. Is it possible to disable that, i.e. to never ask for the password and just fail?

rastytheamateur commented:
> Zděněk wrote something about a prompt for a password. Is it possible to disable that, i.e. to never ask for the password and just fail?

I am trying to figure it out. As far as I know, it should be possible with `GIT_TERMINAL_PROMPT=0` environ variable. Now there is a question how to pass this thing into the python-git-clone-stuff...

Commit:
Ticket #29340 - Add repository check to the signal handler

Commit:
Ticket #29340 - Add branch name into repository str method

Commit:
Ticket #29340 - Use repository branch for cloning

zbohm commented:
[Repo.clone_from](https://github.com/CZ-NIC/djangocms-git-md-page/blob/main/git_md_page/signals/git_update.py#L40) calls [git.clone](https://github.com/gitpython-developers/GitPython/blob/main/git/repo/base.py#L1038) that opens two prompts `Username for 'https://github.com':` and `Password for 'https://user@github.com':`. If credetials are not valid, the `UpdateSignalError` is risen:
```
Cmd('git') failed due to: exit code(128)
  cmdline: git clone -v --depth=1 https://github.com/CZ-NIC/djangocms-oidcxx /tmp/tmppcn647dd
  stderr: 'Cloning into '/tmp/tmppcn647dd'...
remote: Invalid username or password.
fatal: Authentication failed for 'https://github.com/CZ-NIC/djangocms-oidcx/'
```

rastytheamateur commented:
All right, my changes were following:

* Added `branch` attribute into `Repository.__str__` method
* Added params to the `Repo.clone_from` for disabling authentication (or prompts at all, to be exact)
* Forcing git to clone only one single branch to optimize the network load, speed
* In case of a failure during the cloning, set the `content` of all files in the repository to `Repository could not be cloned!`

Question for now is if this solution is good enough. Problem is that the form does not inform the user about possible problems during cloning repositories and searching files. It just alters the `content` attribute.

rastytheamateur commented:
On the other hand, if a repository/branch/file gets deleted, the plugin-admin is not warned neither. I think we do not have to fiddle with form messages and just leave all the responsibility on the plugin-admin who is supposed to ensure that everything is well configured and exists (or will have existed when the page with a plugin instance is published) and will exists as along as the content from the plugin is expected.

zbohm commented:
It is necessary to suppress the prompt. It is also necessary to catch an exception `UpdateSignalError` in `GitTextPluginModel.save`. After that, the server error will no longer be generated and the plugin can be used.

zbohm commented:
Something like this:

```diff
diff --git a/git_md_page/models/git_plugins.py b/git_md_page/models/git_plugins.py
index 70e6240..0a1cab8 100644
--- a/git_md_page/models/git_plugins.py
+++ b/git_md_page/models/git_plugins.py
@@ -42,10 +42,13 @@ class GitTextPluginModel(CMSPlugin):
 
     def save(self, no_signals=False, *args, **kwargs):
         """Send a signal to fetch content."""
-        from git_md_page.signals.git_update import repo_update
+        from git_md_page.signals.git_update import repo_update, UpdateSignalError
 
         signal_sent = kwargs.pop("signal_sent", False)
         super().save(no_signals=no_signals, *args, **kwargs)
         if not signal_sent:
-            # We are triggering a save in the signal, so do not fire again
-            repo_update.send(sender=self.__class__, instance=self)
+            try:
+                # We are triggering a save in the signal, so do not fire again
+                repo_update.send(sender=self.__class__, instance=self)
+            except UpdateSignalError as err:
+                pass
diff --git a/git_md_page/signals/git_update.py b/git_md_page/signals/git_update.py
index 2b92551..fd5d6e1 100644
--- a/git_md_page/signals/git_update.py
+++ b/git_md_page/signals/git_update.py
@@ -37,7 +37,7 @@ def repository_update(sender, **kwargs):
     # Do the update, update content
     temp_folder = mkdtemp()
     try:
-        Repo.clone_from(repository.URL, temp_folder, depth=1)
+        Repo.clone_from(repository.URL, temp_folder, depth=1, env={'GIT_TERMINAL_PROMPT': '0'})
     except CommandError as error:
         raise UpdateSignalError(error)
     else:
```

zbohm commented:
I think cloning a repository after saving an instance is wrong. This should be done in a form clean. In this case the error could be shown on the form.

rastytheamateur commented:
> It is necessary to suppress the prompt. It is also necessary to catch an exception UpdateSignalError in GitTextPluginModel.save. After that, the server error will no longer be generated and the plugin can be used.

Agree, I am going to fix it.

> I think cloning a repository after saving an instance is wrong. This should be done in a form clean. In this case the error could be shown on the form.

I do not really agree. You are right that it would be nice to check everything in the form clean, **but** it would make a bigger sense to check repository (branch) existence in the repository form, not in the file form. On the other side, the plugin instances get updated every time a github web hook is called. So the values of the plugin instances get updated completely independently during the runtime and therefore I consider actual solution (set the `content` to an error message, like `File does not exist.`) as the better one.

jiri-janousek commented:
Since this plugin is long overdue, I'd recommend focusing on MVP.

As far I understand, cloning the repository in Form clean would require a significant rewrite of the plugin (Daniel can elaborate more on that) and doesn't add many benefits as the content of the repository can change anytime.

It would also remove the possibility to include the plugin in advance of the creation of the file in the repository. We would need to discuss this change with Euro-IX whether it is ok for them.

Please try to reach an agreement by Monday's meeting or I'll ask TDi to resolve the issue. I need Daniel's time to work on more important tasks.

zbohm commented:
"This should be done in a form clean." - Of course, I meant the repository form with the repository URL. Not plugin with the branch. Raising an exception and then catching it immediately without any warning to the user, not make any sense at all.

zbohm commented:
MVP is do not block uwsgi processes (suppress prompt) and avoid server error (catch exeption).

zbohm commented:
It's outside the MVP, but have you considered local links? These will not be valid without modification.

jiri-janousek commented:
> It's outside the MVP, but have you considered local links? These will not be valid without modification.

Local links and local images won't work. The plugin was intended to render independent text-only pages as required by Euro-IX. It is not designed to render closely related interlinked pages. It is essential to deliver MVP to show the internal CZ.NIC customers what the plugin from Euro-IX can do and then we can decide whether to improve it further or create a new plugin better suited to rendering the whole subdirectory of a git repository. I got an impression that the customers are unsure about what they need. If we give them MVP, it will be easier for them to specify exact requirements.

jiri-janousek reviewed (commented):
I'm not going to review the code. Zdeněk has provided a lot of useful comments to improve the quality of the plugin and he can review and approve the changes.

Commit:
Ticket #29340 - Add GitRepositoryForm and avoid raising exceptions

rastytheamateur commented:
Last changes:

* Suppressed prompt, so it does not block the process
* Removed raising exceptions, because they were not used anyway, so it does not raise 5xx
* Created admin form for `GitRepository`, which checks only if the repository URL is valid - tries to clone the repository. Branch is not checked since the content can be prepared in advance (or a tag instead of a branch could be possibly used)

Now the most important bugs should be fixed. We can improve the plugin later, the task for now was just to detach it from EuroIX.

zbohm commented on git_md_page/forms/git_plugins.py:
```diff
@@ -1,7 +1,36 @@
+        try:
+            Repo.clone_from(data["URL"], temp_folder, env={"GIT_TERMINAL_PROMPT": "0"}, depth=1)
+        except CommandError:
+            self.add_error("URL", _("Repository could not be cloned."))
```
The real cause of the problem should not be hidden. The message, that it is impossible to clone, is probably useless.

rastytheamateur replied:
Agree, but

```
Cmd('git') failed due to: exit code(128)
  cmdline: git clone -v --depth=1 https://github.com/CZ-NIC/djangocms-git-md-page2.git /tmp/tmpikhbuetd
  stderr: 'Cloning into '/tmp/tmpikhbuetd'...
fatal: could not read Username for 'https://github.com': terminal prompts disabled
```

is not really user friendly. The last line is continuation of the previous `stderr` output, it is not possible to receive that part from an attribute from the `CommandError`. The only thing I am able to get is the error code, `128`. I can add it into the error message, but it will not help to the user.

zbohm replied:
I suggest displaying the whole error this way:
```python
except CommandError as error:
    self.add_error("URL", str(error))
```

rastytheamateur replied:
Fixed.

zbohm commented on git_md_page/signals/git_update.py:
```diff
@@ -27,23 +23,32 @@ def repository_update(sender, **kwargs):
-        Repo.clone_from(repository.URL, temp_folder, depth=1)
-    except CommandError as error:
-        raise UpdateSignalError(error)
+        Repo.clone_from(
+            repository.URL,
+            temp_folder,
+            env={"GIT_TERMINAL_PROMPT": "0"},
+            multi_options=["--single-branch", "--branch {}".format(repository.branch)],
+            depth=1,
+        )
+    except CommandError:
```
I think that once cloning has taken place (at form clone), it would not have to be cloned a second time.

rastytheamateur replied:
But it took place at the `GitRepository` form, not `GitTextPlugin` form.

zbohm commented:
OK for me. It can be merged.
